scripts/imgtool/image.py

```diff
@@ -307,3 +309,48 @@ def pad_to(self, size):
         pbytes += b'\xff' * (tsize - len(boot_magic))
         pbytes += boot_magic
         self.payload += pbytes
+
+    @staticmethod
+    def verify(imgfile, key):
+        with open(imgfile, "rb") as f:
+            b = f.read()
+
+        magic, _, header_size, _, img_size = struct.unpack('IIHHI', b[:16])
+        if magic != IMAGE_MAGIC:
+            raise Exception("Invalid image magic; is this a MCUBoot image?")
+
+        tlv_info = b[header_size+img_size:header_size+img_size+TLV_INFO_SIZE]
+        magic, tlv_tot = struct.unpack('HH', tlv_info)
+        if magic != TLV_INFO_MAGIC:
+            raise Exception("Invalid TLV magic; is this a MCUBoot image?")
+
+        sha = hashlib.sha256()
+        sha.update(b[:header_size+img_size])
+        digest = sha.digest()
+
+        tlv_off = header_size + img_size
+        tlv_end = tlv_off + tlv_tot
+        tlv_off += TLV_INFO_SIZE  # skip tlv info
+        while tlv_off < tlv_end:
+            tlv = b[tlv_off:tlv_off+TLV_SIZE]
+            tlv_type, _, tlv_len = struct.unpack('BBH', tlv)
+            if tlv_type == TLV_VALUES["SHA256"]:
+                off = tlv_off + TLV_SIZE
+                if digest == b[off:off+tlv_len]:
+                    print("sha256sum matched!")
+                    if key is None:
+                        return
+                else:
+                    raise Exception("Invalid hash for image")
+            elif key is not None and tlv_type == TLV_VALUES[key.sig_tlv()]:
+                off = tlv_off + TLV_SIZE
+                tlv_sig = b[off:off+tlv_len]
+                payload = b[:header_size+img_size]
+                try:
+                    key.verify(tlv_sig, payload)
+                    print("signature validated!")
+                    return
+                except InvalidSignature:
+                    raise Exception("Invalid signature for image")
+            tlv_off += TLV_SIZE + tlv_len
+        raise Exception("Signature not found")
```

Comment:
I wonder if we should be pickier about what is in the TLVs? Also, does this work if there are multiple signatures (say with different signers)? It could check that at least one key is valid instead of failing on the first key.

Reply:
> It could check that at least one key is valid instead of failing on the first key.

I'll fix this.

Reply:
Updated to not fail on the first key that didn't match and just printing the exception on the main call, instead of throwing the exception on the user. Also updated commit message to mention that encrypted messages are not yet supported.